### PR TITLE
RFC: ECMA-402 API surface for rust.

### PR DESCRIPTION
**Pull request by filmil**
This is work in progress, still.

**SAMPLE API REVIEW**

### CODE REVIEW

**Comment by filmil** [comment]
This is an example RFC for the API.

**Comment by zbraniecki** [comment]
Interesting. So your trait proposal doesn't own the data!
`unic_langid` *does* own the data.

What's your reasoning behind not owning it?

**Comment by zbraniecki** [comment]
Two more differences:
 - Your proposal accepts subtags via some option bag, `unic_langid` uses `set_region`, `set_script` for that.
 - Your proposal conflates language identifier and locale. I'm inclined to promote the separation.

**Comment by sffc** [comment]
Ownership is an interesting question here.  Is there a way to define the trait in such a way that the implementation can choose whether or not to own the strings?

I also like your "array of enums" suggestion for options delivery.  I added that as an alternative to my doc.

**Comment by zbraniecki** [comment]
>  Is there a way to define the trait in such a way that the implementation can choose whether or not to own the strings?

Possibly, yes. @Manishearth may have more here, but I know that for `FluentBundle` we define it as generic over `Borrow<FluentResource>` which allows us to store either `FluentResource` or `&FluentResource` or `Rc<FluentResource>`.

For `Intl.Locale` I'm a bit less certain if its worth it. I struggle to really see a scenario where storing Strings and using the API to just view into them is really beneficial.

In my experience you only need a string on the input, and then inside your app you really want to operate on, store, manipulate and manage the semantic version of a locale which is the struct, or a list of them.

**Comment by filmil** [comment]
> Interesting. So your trait proposal doesn't own the data!
> `unic_langid` _does_ own the data.
> 
> What's your reasoning behind not owning it?

I would not read too much into the code as it is today.  It is mostly an orientation exercise for me.  As for what I actually wanted, it was:

1. Aas @sffc notes, what I actually would like to achieve, is to leave it to the implementers to determine if they want owned or unowned data.  And as you note, for this particular API, it probably doesn't matter but I wanted to prototype quickly how that would look like.
2. I wanted to see what would it take to express unowned data, because in my (limited) experience with rust so far, working with unowned data seems more restrictive. 
3. I wanted to propose a way to organize a discussion around a RFC with actual code.
4. As to how the API should actually look like, I'd probably want to work off of existing concrete code and factor the traits out of those, instead of trying to work out from first principles something that is general enough.

Thoughts?

**Comment by zbraniecki** [comment]
> Aas @sffc notes, what I actually would like to achieve, is to leave it to the implementers to determine if they want owned or unowned data.

That makes sense. In fact, unless an API has the potential to just store the input string, we still only want a reference on the input anyway, because we'll parse it and store in a different form anyway.

What may be interesting in this case, is to use another trait:

```rust
trait LanguageIdentifier: FromStr {}
```

means that any user has to implement `FromStr` for that trait, which gives us a nice fallible parser for `&str`, which is exactly what we want here I believe.

> As to how the API should actually look like, I'd probably want to work off of existing concrete code and factor the traits out of those, instead of trying to work out from first principles something that is general enough.

Could we start with `unic-langid` master? (it has RFC344 for getter names)

File: rfc/002-ecma-402/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+//!
+//! - https://github.com/rust-lang/rfcs/issues/858
+//! - https://github.com/i18n-concept/rust-discuss/issues/14
+//!
+
+/// `Option` is used to specify overriding options for the constructor methods.
+///
+/// Some of the Opts here could take `enum`s instead, to be less stringly typed.
+pub enum Opt {
+    Script(String),
```
Comment (Manishearth) [inline rfc/002-ecma-402/src/lib.rs:29]: Would it be possible to have multiple of the same kind of option specified? If not, should this be a struct of options?

File: rfc/002-ecma-402/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    Region(String),
+    HourCycle(String),
+    Calendar(String),
+}
+
+/// Implements ECMA-402 `Intl.Locale`.
+///
+/// This is an exercise only.
+pub trait Locale<'a>: Sized {
```
Comment (Manishearth) [inline rfc/002-ecma-402/src/lib.rs:38]: I feel like adding a lifetime here is more trouble than it's worth. But @zbraniecki would be a better person to answer which things are best borrowed.

**Review by Manishearth — COMMENTED** [review]
I feel like for small locale strings we should honestly be okay with allocating. There are things like `TinyStr` which can help avoid allocations here.

**Comment by sffc** [comment]
Is there a well-known trait that would accept either a String or a TinyStr?

**Comment by Manishearth** [comment]
@sffc Accepting `Borrow<str>` would probably do it.

But really I was suggesting we use `TinyStr` and similar optimizations _internally_ to avoid allocations, and then pass in `&str` only.

**Comment by sffc** [comment]
So you're saying that internally, we could have fields like this to support a public method that accepts a `&str`?  Note: this code is untested and probably doesn't compile...

```rust
enum MaybeTinyStr {
    Tiny(TinyStr16),
    Long(String)
};

impl Deref for MaybeTinyStr {
    fn deref(&self) -> &str {
        match self {
            Tiny(value) => value.as_str(),
            Long(value) => value.as_str()
        }
    }
};

impl FromStr for MaybeTinyStr {
    type Err = Error;

    fn from_str(text: &str) -> Result<Self, Self::Err> {
        if (str.len() <= 16) {
            TinyStr16::from_str(str);
        } else {
            String::from_str(str);
        }
    }
};

struct IntlObject {
    MaybeTinyStr value;
};

impl IntlObject {
    fn setValue(&mut self, value: &str): void {
        match MaybeTinyStr::from_str(str) {
            Ok(result) => self.value = value,
            Err(_) => self.value = MaybeTinyStr.from_str("").unwrap()
        };
    }
}
```

**Comment by Manishearth** [comment]
Yep! I suspect most of the string fields will be TinyStr sized anyway.

**Comment by zbraniecki** [comment]
I don't understand. For Unicode Language and Locale Identifier *all* fields will be TinyStr. That's the whole point - TinyStr covers all subtags and extensions and core operations between them.

**Comment by sffc** [comment]
We're talking about the general case of accepting an ASCII-range string into the API.  Rather than dealing with lifetime parameters of `&'a str`, we can accept a `&str` and copy it into internal storage.  For cases where the strings are guaranteed to be short enough, we can use TinyStr directly; otherwise, when the length is not known, we could use something along the lines of TinyStrAuto.

**Comment by Manishearth** [comment]
Presumably this is for region codes and extensions and such? Since script and language tags all fit in TinyStr

**Comment by sffc** [comment]
TinyStrAuto could be used for datetime skeletons, number skeletons, measurement unit identifiers, and other variable-length ASCII-only identifiers.

**Comment by zbraniecki** [comment]
> Presumably this is for region codes and extensions and such? Since script and language tags all fit in TinyStr

region code is a `TinyStr4` and all extensions are either 4 or 8. There's no use of `TinyStr16` in `unic-locale` needed, anywhere.

> TinyStrAuto could be used for datetime skeletons, number skeletons, measurement unit identifiers, and other variable-length ASCII-only identifiers.

I see. Is the goal also to allow for us to introduce `TinyStr32` in the future and add it to `TinyStrAuto` instead of `TinyStr16` in a backward compatible model - is the decision to use string or tinystr an internal detail meant to be opaque for the user?

**Comment by filmil** [comment]
@zbraniecki Is there a reason why `TinyStr__` must be a first-class member of the API, instead of an implementation detail?

**Comment by zbraniecki** [comment]
> Is there a reason why TinyStr__ must be a first-class member of the API, instead of an implementation detail?

Memory layout. For use cases like `unic-langid` its known what size you need, and using some sort of `auto` would be a waste of space.

**Comment by filmil** [comment]
> Memory layout. For use cases like `unic-langid` its known what size you need, and using some sort of `auto` would be a waste of space.

But `TinyStrXX` is not a correctness requirement, right? So should not necessarily be part of the traits, at least IMHO. Do I misunderstand?

**Comment by sffc** [comment]
I never intended for TinyStr to be part of the trait API. I was discussing it to argue that we can accept `&str` in the trait and have somewhere efficient to copy the string, rather than saving a lifetimed reference to the string.

**Comment by zbraniecki** [comment]
Agree. I misread your question. TinyStr should not be part of the trait - it's an implementation detail. The trait should implement `FromStr` and return `&str`. I also would say that for perf reasons a constructor like `from_bytes(v: &[u8]) -> Result<Self, LanguageIdentifierError>` would be great because we still need to validate the input is an ASCII str no matter what.
For example, in Gecko we use `from_bytes` - https://searchfox.org/mozilla-central/source/intl/locale/rust/unic-langid-ffi/src/lib.rs#18

**Comment by Manishearth** [comment]
Right, TinyStr should be internal, and we should accept `&str`s with a simple copy of the bytes into the TinyStr (occasionally allocating if large)

**Comment by sffc** [comment]
I'm closing this PR so that we can move the repo to unicode-org.  Feel free to re-open later, at which point the Unicode CLA will take effect.

**Comment by filmil** [comment]
STGM

**Pull request closed without merging**